SwiftCompilerSources/Sources/Optimizer/Utilities/BorrowUtils.swift

```diff
@@ -0,0 +1,832 @@
+  /// incoming value dominates or is consumed by an outer adjacent
+  /// phi. See InteriorLiveness.
+  ///
+  /// TODO: to hande reborrow-extended uses migrate ExtendedLiveness
+  /// to SwiftCompilerSources.
+  ///
+  /// TODO: For instructions that are not a BeginBorrowValue, verify
+  /// that scope ending instructions exist on all paths. These
+  /// instructions should be complete after SILGen and never cloned to
+  /// produce phis.
+  func visitScopeEndingOperands(_ context: Context,
+                                visitor: @escaping (Operand) -> WalkResult)
+  -> WalkResult {
+    switch self {
+    case .beginBorrow, .storeBorrow:
+      let svi = instruction as! SingleValueInstruction
+      return svi.uses.filterUsers(ofType: EndBorrowInst.self).walk {
+        visitor($0)
+      }
+    case .beginApply(let bai):
+      return bai.token.uses.walk { return visitor($0) }
+    case .partialApply(let pai):
+      return visitForwardedUses(introducer: pai, context) {
+        switch $0 {
+        case let .operand(operand):
+          if operand.endsLifetime {
+            return visitor(operand)
+          }
+          return .continueWalk
+        case let .deadValue(_, operand):
+          if let operand = operand {
+            assert(!operand.endsLifetime,
+                   "a dead forwarding instruction cannot end a lifetime")
+          }
+          return .continueWalk
+        }
+      }
+    case .startAsyncLet(let builtin):
+      return builtin.uses.walk {
+        if let builtinUser = $0.instruction as? BuiltinInst,
+          builtinUser.id == .EndAsyncLetLifetime {
+          return visitor($0)
+        }
+        return .continueWalk
+      }
+    }
+  }
+
+  var description: String { instruction.description }
+}
+
+/// A value that introduces a borrow scope:
+/// begin_borrow, load_borrow, reborrow, guaranteed function argument.
+///
+/// If the value introduces a local scope, then that scope is
+/// terminated by scope ending operands. Function arguments do not
+/// introduce a local scope because the caller owns the scope.
+///
+/// If the value is a begin_apply result, then it may be the token or
+/// one of the yielded values. In any case, the scope ending operands
+/// are on the end_apply or abort_apply intructions that use the
+/// token.
+///
+/// Note: equivalent to C++ BorrowedValue, but also handles begin_apply.
+enum BeginBorrowValue {
+  case beginBorrow(BeginBorrowInst)
+  case loadBorrow(LoadBorrowInst)
+  case beginApply(Value)
+  case functionArgument(FunctionArgument)
+  case reborrow(Phi)
+
+  init?(_ value: Value) {
+    switch value {
+    case let bbi as BeginBorrowInst: self = .beginBorrow(bbi)
+    case let lbi as LoadBorrowInst: self = .loadBorrow(lbi)
+    case let arg as FunctionArgument: self = .functionArgument(arg)
+    case let arg as Argument where arg.isReborrow:
+      self = .reborrow(Phi(arg)!)
+    default:
+      if value.definingInstruction is BeginApplyInst {
```

Comment:
nit, in case you are pushing a new version anyway: you can write
```
  case let mvir as MultipleValueInstructionResult where mvir.parentInstruction is BeginApplyInst:
```
Then you can also define the case payload as `case beginApply(MultipleValueInstructionResult)` which avoids the `!` when getting its parent instruction.